### ISSUE TITLE
Unable to retrieve next page of products from fetchNextPage

### ISSUE THREAD

**Issue by Nhawdge** [Shopify/js-buy-sdk#472]
I've got an issue retrieving the next page of products. It's similar to this [comment](https://github.com/Shopify/js-buy-sdk/issues/433#issuecomment-353181172) on issue #433. Here is my code (I've removed the unimportant bits).

        ShopifyClient.collection.fetchWithProducts(categoryId).then(function (collection) {
            console.log("Loading Collecton: ", collection);
            for (var product of collection.products) {                
                // Handling collection
                lastProduct = product;
            }

            function LoadNextPage(lastProduct) {
                console.log("Get next page", lastProduct);
                self.ShopifyClient.fetchNextPage(lastProduct).then(function (nextPageOfProducts) {
                    console.log(nextPageOfProducts);
                    for (var product of nextPageOfProducts.model) {
                        // Handling collection
                        lastProduct = product;
                    }
                    LoadNextPage(lastProduct);
                })
            };

            LoadNextPage(lastProduct);

        }).catch(function (ex) {
            console.error(ex);
        })

This code seems to work fine for the first page, the second page, and the third page, however when I attempt to retrieve the fourth page I receive.

```
shopify-buy.umd.polyfilled.min.js:25 Uncaught (in promise) TypeError: Cannot read property 'node' of undefined
    at shopify-buy.umd.polyfilled.min.js:25
    at Array.reduce (<anonymous>)
    at shopify-buy.umd.polyfilled.min.js:25
    at <anonymous>
```

I am using what I believe is v1.0.2 I cannot find anything stating its version, however, it is referenced [here](https://shopify.github.io/js-buy-sdk/) as 
```
<script src="http://sdks.shopifycdn.com/js-buy-sdk/v1/latest/index.umd.min.js"></script>
```

I was using the collection as the argument for my `LoadNextPage` function, however that failed immediately with a very similar error.

Thanks for your help!

**Comment by minasmart** [comment]
I'm guessing you're trying to fetch beyond the last page.

A couple notes: you can just do `fetchNextPage(collection.products)`. The function will automatically use the cursor from  the last element of the array.

Next you should probably check `collection.products.hasNextPage` before calling `fetchNextPage` on the array. Does this help?

**Comment by Nhawdge** [comment]
Hi @minasmart Thanks for the fast reply! I rewrote my code based on your suggestion to use the collection, The part that makes it a little tricky is that the first response gives a category with a list of products (`category.products`), and the `fetchNextPage()` gives an object with a data array of products, (`nextPage.data`);  Here is the updated code, again with the excess removed. I also have an output of my console to hopefully share a better idea of the issue I'm experiencing.

        self.ShopifyClient.collection.fetchWithProducts(categoryId).then(function (collection) {
            console.log("Loading Collecton: ", collection);
            for (var product of collection.products) {
                 // My handling
             }

            function LoadNextPage(lastProductList) {
                console.log("Last Page: ", lastProductList);
                self.ShopifyClient.fetchNextPage(lastProductList).then(function (nextPageOfProducts) {
                    console.log("Next Product List: ", nextPageOfProducts);
                    for (var product of nextPageOfProducts.model) {
                        // My handling
                    }
                    LoadNextPage(nextPageOfProducts.model);
                }).catch(function (error) { console.error(error); })
            };

            LoadNextPage(collection.products);

        }).catch(function (ex) {
            console.error(ex);
        })

Using this code to fetch my collection retreives page 1, 2, and 3, but not 4. ( I expanded the last item in the last successful response but deleted the excess for the sake of reading)

```
Loading Collecton:  e {…}description: (...)descriptionHtml: (...)handle: (...)id: (...)image: (...)products: (...)title: (...)type: (...)updatedAt: (...)__typename: (...)attrs: {__typename: e, id: e, handle: e, description: e, descriptionHtml: e, …}get description: ƒ ()get descriptionHtml: ƒ ()get handle: ƒ ()get id: ƒ ()get image: ƒ ()get products: ƒ ()get title: ƒ ()get type: ƒ ()get updatedAt: ƒ ()get __typename: ƒ ()__proto__: Object

14:26:36.571 ShopViewModel.js:208 Last Page:  (20) [e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, type: {…}]0: e {…}1: e {…}2: e {…}3: e {…}4: e {…}5: e {…}6: e {…}7: e {…}8: e {…}9: e {…}10: e {…}11: e {…}12: e {…}13: e {…}14: e {…}15: e {…}16: e {…}17: e {…}18: e {…}19: e {…}type: {name: "ProductConnection", kind: "OBJECT", fieldBaseTypes: {…}, implementsNode: false}length: 20__proto__: Array(0)

14:26:37.237 ShopViewModel.js:210 Next Product List:  {data: {…}, model: Array(20)}

14:26:37.282 ShopViewModel.js:208 Last Page:  (20) [e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, type: {…}]

14:26:38.327 ShopViewModel.js:210 Next Product List:  {data: {…}, model: Array(20)}

14:26:38.404 ShopViewModel.js:208 Last Page:  (20) [e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, e, type: {…}]
19: 
  hasNextPage: true
  hasPreviousPage: e {value: true}

14:26:38.584 ShopViewModel.js:232 TypeError: Cannot read property 'node' of undefined
    at shopify-buy.umd.polyfilled.min.js:25
    at Array.reduce (<anonymous>)
    at shopify-buy.umd.polyfilled.min.js:25
    at <anonymous>
```

I appreciate your help!

**Comment by minasmart** [comment]
On the request that's failing, if you check in your network inspector, there should be a header returned called `X-Request-ID`. Could you pass that along? From any of the failed requests? This may be an error on our side, but I need more detail.

Secondly, because we're using a graphql API, the next pagination will never  be structured like the original set. We can't request next pages through a million object graph layers, so we pear down the tree and just return the next page. The underlying graphql client returns a hash of a `model` and `data`, `data` is the raw response, `model` is the rich data from the graphql client. Hopefully this explains some things. If I can get the request ID, I can probably see more about what's going on here.

**Comment by Nhawdge** [comment]
Here's my last request header prior to failing: `X-Request-ID:50e02b09-84d6-4849-b5d2-69711d4ea641`

Here is a request header that was successful for me: `X-Request-ID:5bb34d0b-9832-43e2-9823-3b9d8ef1bab9`

I also appreciate the explanation, this is my first time working with graphql. Would it be beneficial for me to read into that API more, or is that more on your side of it? 

Thanks!

**Comment by Nhawdge** [comment]
@minasmart Just curious if there was any update or a work around I can do for the time being? We're trying to meet an upcoming deadline on our project that depends on this API.

Thank you!

**Comment by Nhawdge** [comment]
Hello, I am just looking to see if there is any status update I can pass along to the plus store we are working with for this issue.

Thanks!

**Comment by minasmart** [comment]
I haven't had time to look at this over the last few days. I'll hopefully have a chance this afternoon or next week.

**Comment by Nhawdge** [comment]
Good afternoon, Just looking to follow up on this issue to see if there has been any findings I can share with my client on this issue. 

Thank you!

**Comment by Nhawdge** [comment]
@minasmart  I've been doing a large amount of digging into this issue and I found the cause, however I don't know how this is written well enough to track down the actual cause. 

In the `ShopifyClient.send` function, variable `graphQLParams`  had a dollar sign character in an inappropriate place.  The true error message was `"Parse error on "$" (VAR_SIGN) at [1, 573]"
` and caused by this portion of the query `first: $first` 

I cannot find the source of `$first` anywhere in the object. I have a very hacky solution below and am hoping you'd be able to track down the source and with a cleaner approach.

I modified line 2001 to be 
```
var graphQLParams = { query: operationOrDocument.toString().replace(/\$first\)/, '20)') };
```

I am not sure why it is only my application seeing this issue, but I hope it helps!
